### PR TITLE
:running: Bump k8s.io/* deps to v0.31.0-alpha.3

### PR DESCRIPTION
**Pull request by twz123**
Adapt to upstream changes:

* kubernetes/kubernetes#124917
  Wrap the anonymous flag in a struct pointer.

* kubernetes/kubernetes#125560
  Add `CreateOptions` and `UpdateOptions` to the existing `versionedTracker` methods, and also add the `Patch` method that was introduced upstream.

### CODE REVIEW

**Comment by sbueringer** [comment]
@twz123 Can you please split up the changes in separate PRs if possible? I assume this change "https://github.com/kubernetes/kubernetes/pull/125560" can be done in a separate PR?

**Comment by twz123** [comment]
> I assume the follow-up to this change [kubernetes/kubernetes#125560](https://github.com/kubernetes/kubernetes/pull/125560) can be done in a separate PR?

The addition of the variadic options arguments is necessary to make it compile, and the addition of the Patch method is necessary to pass the fake client unit tests, as `k8s.io/client-go/testing.ObjectReaction` changed the way it reacts to Patch actions. I can split up the fixes for the tests into a separate PR, if you prefer.

**Comment by sbueringer** [comment]
Ah thx for the context. No I think then let's keep everything in this one PR

File: pkg/client/fake/client.go [hunk 1]
```diff
@@ -359,66 +359,101 @@ func convertFromUnstructuredIfNecessary(s *runtime.Scheme, o runtime.Object) (ru
...
+
+	return t.update(gvr, obj, ns, false, false, updateOpts)
+}
+
+func (t versionedTracker) update(gvr schema.GroupVersionResource, obj runtime.Object, ns string, isStatus, deleting bool, opts metav1.UpdateOptions) error {
+	obj, err := t.updateObject(gvr, obj, ns, isStatus, deleting, opts.DryRun)
+	if err != nil {
+		return err
+	}
+	if obj == nil {
```
[the remark this answers is not in the document]
Comment (twz123) [reply pkg/client/fake/client.go:376]: `updateObject` can return `nil, nil` currently, if an object is created ([line 430](https://github.com/kubernetes-sigs/controller-runtime/pull/2872/commits/b068c2cef08fa3c49581573b4618fb21622843ad#diff-20ecedbf30721c01c33fb67d911da11c277e29990497a600d20cb0ec7215affdR430)) or deleted ([line 491](https://github.com/kubernetes-sigs/controller-runtime/pull/2872/commits/b068c2cef08fa3c49581573b4618fb21622843ad#diff-20ecedbf30721c01c33fb67d911da11c277e29990497a600d20cb0ec7215affdR491)). In that case, the object tracker shouldn't be called with Update/Patch.

**Review by alvaroaleman — APPROVED** [review]
thanks!

**Pull request merged**